### PR TITLE
Suggestion to add exact sum functions (call accumulate function) in StatisticsFunctions folder

### PR DESCRIPTION
**Pull request by Silfurion**
Hello,

In some of our projects, we would like to accelerate the sum of vector elements. At the moment CMSIS doesn't offer a simple sum function. The closest method was the mean method that unfortunately doesn't offer an exact result if we want to multiply its result by the size of the array to get back the sum of the vector.

With this pull request we would like to suggest the addition of a new sum function that would allow to benefit from the already implemented and accelerated mean function, but without the final division to obtain a more accurate sum result.

We added these function in the StatisticsFunctions folder available for q7 , q15 , q31 , float16 , float32 and float64.

Thanks again

### CODE REVIEW

**Comment by christophe0606** [comment]
@Silfurion @jbr-smtg There are some problems with the fixed point arithmetic versions. I am going to add some comments to the files in the commit explain what I mean.

**Comment by Silfurion** [comment]
Hello,

I didn't find the comments you mentioned in the last message (I may have searched wrong). I wanted to know where you put these comments so I could make the changes.

I apologize for the inconvenience and thank you in advance for your response.

**Comment by christophe0606** [comment]
@Silfurion Hello,

Just look above on this page.

You'll see code extracts and my comments below.
Or you can also look at the "Files changes" tab and look for the comments.

I know you are mainly working with float in your team. Here the problems are only related to fixed point arithmetic. If you want to push only floating point versions then it should also be ok at least for this set of functions.

**Review by christophe0606 — CHANGES_REQUESTED** [review]

**Comment by Silfurion** [comment]
Hello, 

I changed the output type so as to avoid the saturation. For the q7 version I used a q31 output type and for other version (q15 and q31) I used the q63 type. Now I would like to change the comments but I don't understand the logic behind the format (like 17.15 format). Do you have resources about that or can you explain to me (if you have time of course) this logic.

I thank you in advance and wish you a good day.

File: Source/StatisticsFunctions/arm_accumulate_q15.c [hunk 1]
```diff
@@ -57,11 +57,11 @@
 void arm_accumulate_q15(
                         const q15_t * pSrc,
                         uint32_t blockSize,
-                        q15_t * pResult)
+                        q63_t * pResult)
 {
   uint32_t  blkCnt;           /* loop counters */
   q15x8_t  vecSrc;
-  q31_t     sum = 0L;
+  q63_t     sum = 0L;
```
Comment (christophe0606) [inline Source/StatisticsFunctions/arm_accumulate_q15.c:64]: This is not consistent with the MVE instruction used : vaddvaq is using a int32.
So either you accumulate on int32 only. Or you need to internally expand the values to int32 and then use the MVE instruction to accumulate on int64.

**Review by christophe0606 — CHANGES_REQUESTED** [review]
I think you should limit your PR to things you can build and test.

The fixed point arithmetic part has too many open questions and / or problems and I don't think you have attempted to build and run this part on a M55 using MVE instruction set.

I suggest you limit this PR to float only and remove the q7, q15, q31 parts ...

Regarding the float there is still an open question : you can do it in a simple way (like other part of the CMSIS-DSP where we internally also do some accumulation) or you can use Kahan summation for better accuracy (but less speed). What is the level of accuracy you need for your application ? 

If you need Kahan summation then I suggest to make another PR with the more accurate implementation so that people who don't need it can use the faster (but less accurate) implementation provided in this PR.

**Review by christophe0606 — APPROVED** [review]
Thanks. I'll merge it when I am back from vacations. I don't want to merge anything just before leaving. And also, I'll have to write some tests for the function.

**Comment by christophe0606** [comment]
@Silfurion I have written the tests and merged the PR.
Next time, please add #if defined(__aarch64__) when the code can't be built on Arm core with Neon but without aarch64 support.

I have not tested the aarch64 part since, so far, it was not a target of CMSIS-DSP.
I'll try to add this to my test framework. So, validation if this PR is not fully done and I won't start on the other ones before I have finished this one.

**Pull request closed without merging**